### DISCUSSION TITLE
can be pydot used instead of PyGraphviz ?

### DISCUSSION THREAD

**Discussion by Sandy4321** [networkx/networkx#5365] [Q&A]
it is written 
[PyGraphviz](http://pygraphviz.github.io/) and [pydot](https://github.com/erocarrera/pydot) provide graph drawing and graph layout algorithms via [GraphViz](http://graphviz.org/).

but PyGraphviz still have problems with installation when 
pydot maybe installed correctly 

may you help to understand 
1
in which capacity PyGraphviz is needed ?
2
can be pydot used instead of PyGraphviz ?

**Comment by rossbar** [comment]
There are several questions/comments here, so I'll try to address them individually. A caveat: I am not a Graphviz expert nor have I ever used `pydot`, so my perspective is quite limited. That said:

> but PyGraphviz still have problems with installation when pydot maybe installed correctly

Installation of PyGraphviz should be vastly improved in the latest version (1.9) over earlier versions, but yes - the trick here is that you must have Graphviz installed on your system in order to be able to install pygraphviz. Probably the simplest way to do this (on Windows at least) is to use `conda`, though **make sure you are installing from the conda-forge channel**. See the `pygraphviz` [installation docs](https://pygraphviz.github.io/documentation/stable/install.html) for more info.

`pydot` is a pure-Python package, so you shouldn't have any trouble installing, though there have recently been some issues with the newest versions of `pyparsing` which have affected `nx_pydot`. See e.g. pydot/pydot#277

> it is written [PyGraphviz](http://pygraphviz.github.io/) and [pydot](https://github.com/erocarrera/pydot) provide graph drawing and graph layout algorithms via [GraphViz](http://graphviz.org/).

`pygraphviz` wraps Graphviz thus Graphviz is a hard dependency, whereas pydot is (AIUI) a standalone library that translates directly to/from Graphviz's `.dot` format. Therefore Graphviz is a soft dependency for `pydot` - if all you need to do is read/write `.dot` files, `pydot` should be perfectly adequate. However, if you want to e.g. use Graphviz's layout functions, you will need Graphviz to be installed, even in the `pydot` case.

From a user perspective, the main difference between the two libraries is API. PyGraphviz has a more networkx-like API:

```python
>>> import pygraphviz as pgv
>>> A = pgv.AGraph()
>>> A.add_edges_from([(0, 1), (1, 2)])
>>> A.edges
[('0', '1'), ('1', '2')]
```

The above should look familiar to any NetworkX user. `pydot` provides it's own API for creating/modifying graphs (see `pydot.Dot`). NetworkX supplies explicit conversion functions for both libraries.

>  in which capacity PyGraphviz is needed ?

`pygraphviz` is intended to be a complete wrapper for Graphviz, meaning that in principle you *should* be able to use any/all Graphviz features in Python via pygraphviz.

>  can be pydot used instead of PyGraphviz ?

Yes, though it depends what you want to do. If you are interested mostly in reading/writing `.dot` files, then `pydot` should be completely sufficient (even preferred if it is difficult to install Graphviz on your system). AIUI rendering the `.dot` files to images will require Graphviz.

There are of course many other details, but hopefully this overview helps!

**Comment by Sandy4321** [comment]
thanks for soon answer 
I meant to make plots using pydot
